### PR TITLE
Manager policy mapping first draft

### PR DESCRIPTION
**Pull request by larkinscarvalho**

### CODE REVIEW

File: designs/manager_policy_mapping/policy_mapping.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
+# Policy mapping workflow with IP/Mac
```
Comment (emanoelxavier) [inline designs/manager_policy_mapping/policy_mapping.md]: picky: Policy Mapping Workflow with IP/MAC

File: designs/manager_policy_mapping/policy_mapping.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+This document describes the proposed changes for policy mapping workflow with IP/Mac.
```
Comment (emanoelxavier) [inline designs/manager_policy_mapping/policy_mapping.md]: IPs/MACs

File: designs/manager_policy_mapping/policy_mapping.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+## Background
+
+Currently, we have two ways to determine which policy gets applied to which workload:
+
+- Provide manager with encapsulation tag
```
Comment (emanoelxavier) [inline designs/manager_policy_mapping/policy_mapping.md]: manager -> the VNF security manager

File: designs/manager_policy_mapping/policy_mapping.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+- Provide manager with IP/Mac
+
+If the manager is supporting policy mapping and security group, we provide manager security group information to bind the policy.
```
Comment (emanoelxavier) [inline designs/manager_policy_mapping/policy_mapping.md]: Mention the name of the properties here referred to in the plugin documentation and/or provide a link to the line of code.  
We provide manager security group -> we provide the manager with the security group  name and members (their IPs and MACs)

File: designs/manager_policy_mapping/policy_mapping.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+## Constraints and Assumptions
+The Security Group on the manager with one or more policies/interfaces is never deleted.
```
Comment (emanoelxavier) [inline designs/manager_policy_mapping/policy_mapping.md]: The appliance security manager **will not delete** a security group with one or more security group interfaces.

File: designs/manager_policy_mapping/policy_mapping.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+## Design Changes
+`SecurityGroup`: Currently, `SecurityGroup` and `SecurityGroupInterface` have many to many entity mapping. The proposed design will have one to many mapping of `SecurityGroup` with `SecurityGroupInterface`.
```
Comment (emanoelxavier) [inline designs/manager_policy_mapping/policy_mapping.md]: will change this relationship to a one to many relationship between the SG and the SGIs

File: designs/manager_policy_mapping/policy_mapping.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+### REST API
+`SecurityGroupInterfaceDto`:
+
+```
+[
+  {
+    "virtualSystemId": 0,
+    "name": "string",
+    "policyId": 0,
+    "mgrSGId": "string",        //New API property
```
Comment (emanoelxavier) [inline designs/manager_policy_mapping/policy_mapping.md]: Can you make it bold or not possible? Also make a full statement to make it easier to spot // This new property is being added to store the security group id returned by the manager

File: designs/manager_policy_mapping/policy_mapping.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        "mgrPolicyId": "string",
+        "mgrDomainId": 0,
+        "mgrDomainName": "string"
+      }
+    ],
+    "markedForDeletion": false,
+    "binded": false
+  }
+]
+```
```
Comment (emanoelxavier) [inline designs/manager_policy_mapping/policy_mapping.md]: You are also changing the the securitygroupdto right? I thought the SG manager id field was being removed.

File: designs/manager_policy_mapping/policy_mapping.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+### OSC SDKs
+
+#### VNF Security Manager SDK
+
+`SecurityGroupInterfaceInput`: Create package org.osc.sdk.manager.input and add the `SecurityGroupInterfaceInput` interface with `name`, `mgrSGId` and `tag` of type string.
```
Comment (emanoelxavier) [inline designs/manager_policy_mapping/policy_mapping.md]: InSecurityGroupInterface and OutSecurityGroupInterface  (for return types), @arvindn05 what you think? 
Larkins

File: designs/manager_policy_mapping/policy_mapping.md [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+`ManagerSecurityGroupInterfaceApi`: Add `SecurityGroupInterfaceInput` as a paramater to interface methods `createSecurityGroupInterface` and `updateSecurityGroupInterface`
+
+#### SDN Controller SDK
+No changes to SDN Controller SDK are required.
+
+### XYZ Integration Point
+
```
Comment (emanoelxavier) [inline designs/manager_policy_mapping/policy_mapping.md]: Call out this represents a breaking change to the other plugins. and you are updating the one below

File: designs/manager_policy_mapping/policy_mapping.md [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+`security-mgr-sample-plugin`: Update sample manager plugin to incorporate new changes to the API's
+
+### OSC UI
+No changes to the OSC UI are required.
+
+### OSC Entities
+`SecurityGroupInterface`: Add `mgrSGId` entity of type string
```
Comment (emanoelxavier) [inline designs/manager_policy_mapping/policy_mapping.md]: not entity, field

File: designs/manager_policy_mapping/policy_mapping.md [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+`SecurityGroup`: Delete the `mgrId` entity
```
Comment (emanoelxavier) [inline designs/manager_policy_mapping/policy_mapping.md]: not entity, field

File: designs/manager_policy_mapping/policy_mapping.md [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+### OSC Synchronization Tasks
+Update tasks affected by the API changes
```
Comment (emanoelxavier) [inline designs/manager_policy_mapping/policy_mapping.md]: You should mention the ones being affected.

File: designs/manager_policy_mapping/policy_mapping.md (hunk 7 above)
Comment (larkinscarvalho) [inline designs/manager_policy_mapping/policy_mapping.md]: The content inside the block code is bold by default.

File: designs/manager_policy_mapping/policy_mapping.md (hunk 8 above)
Comment (larkinscarvalho) [inline designs/manager_policy_mapping/policy_mapping.md]: We are deleting the SG manager id field. However, the SecurityGroupDto is N/A

File: designs/manager_policy_mapping/policy_mapping.md (hunk 9 above)
Comment (larkinscarvalho) [inline designs/manager_policy_mapping/policy_mapping.md]: Waiting for @arvindn05 inputs

File: designs/manager_policy_mapping/policy_mapping.md [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+This document describes the proposed changes for policy mapping workflow with IPs/MACs.
+
+## Background
+
+Currently, we have two ways to determine which policy gets applied to which workload:
+
+- Provide the VNF security manager with encapsulation tag
+- Provide the VNF security manager with IP/MAC
+
+If the manager is supporting [Policy Mapping](https://github.com/opensecuritycontroller/security-mgr-api/blob/master/src/main/java/org/osc/sdk/manager/api/ApplianceManagerApi.java#L40) and [Security Group](https://github.com/opensecuritycontroller/security-mgr-api/blob/master/src/main/java/org/osc/sdk/manager/api/ApplianceManagerApi.java#L41), we provide the manager with the security group name and members (their IPs and MACs) information to bind the policy.
```
Comment (emanoelxavier) [inline designs/manager_policy_mapping/policy_mapping.md]: Use the links to the Const file instead, i.e.: https://github.com/opensecuritycontroller/security-mgr-api/blob/master/src/main/java/org/osc/sdk/manager/Constants.java#L70

File: designs/manager_policy_mapping/policy_mapping.md [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+#### VNF Security Manager SDK
+
+`SecurityGroupInterfaceElement`: Create package org.osc.sdk.manager.input and add the `SecurityGroupInterfaceElement` interface with `name`, `mgrSGId` and `tag` of type string.
+
+`ManagerSecurityGroupInterfaceApi`: Add `SecurityGroupInterfaceElement` as a paramater to interface methods `createSecurityGroupInterface` and `updateSecurityGroupInterface`
+
+#### SDN Controller SDK
+No changes to SDN Controller SDK are required.
+
+### XYZ Integration Point
```
Comment (emanoelxavier) [inline designs/manager_policy_mapping/policy_mapping.md]: This is a placeholder header, please change this to VNF Security Manager Plugins

File: designs/manager_policy_mapping/policy_mapping.md [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+The change to the VNF Security Manager SDK will break the plugin implementations. The plugin authors are required to update their plugins.
```
Comment (emanoelxavier) [inline designs/manager_policy_mapping/policy_mapping.md]: Please change: The plugin authors are required to update their plugins -> Updating the plugins implementing these APIs is out of scope for this feature, with the exception of the `security-mgr-sample-plugin` which will be modified accordingly. 
With this you can remove the ln 62.

File: designs/manager_policy_mapping/policy_mapping.md [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+### REST API
+`SecurityGroupInterfaceDto`:
+
+```
+[
+  {
+    "virtualSystemId": 0,
+    "name": "string",
+    "policyId": 0,
+    "mgrSGId": "string", // This new property is being added to store the security group id returned by the manager
```
Comment (arvindn05) [inline designs/manager_policy_mapping/policy_mapping.md]: mgrSGId -> managerSecurityGroupId

File: designs/manager_policy_mapping/policy_mapping.md [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+    "binded": false
+  }
+]
+```
+
+### OSC SDKs
+
+#### VNF Security Manager SDK
+
+`SecurityGroupInterfaceElement`: Create package org.osc.sdk.manager.input and add the `SecurityGroupInterfaceElement` interface with `name`, `mgrSGId` and `tag` of type string.
```
Comment (arvindn05) [inline designs/manager_policy_mapping/policy_mapping.md]: mgrSGId -> managerSecurityGroupId

Also describe that tag = policy in which cases
and tag = VLAN/Network tag in which cases

I believe if you can add the interface example that would be clear instead of describing the interface. see example @ https://github.com/karimull/community/blob/b1b6377c829ff4fec7aa156b81c9fddfc99f2ed7/designs/service_function_chaining/SFC_design.md#interface-changes

File: designs/manager_policy_mapping/policy_mapping.md [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+### OSC SDKs
+
+#### VNF Security Manager SDK
+
+`SecurityGroupInterfaceElement`: Add a new interface
+
+```
+public interface SecurityGroupInterfaceElement {
+
+    /**
```
Comment (arvindn05) [inline designs/manager_policy_mapping/policy_mapping.md]: please fix formatting

File: designs/manager_policy_mapping/policy_mapping.md [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+	 * Name of the security group
+	 */
+    String getName();
+
+	/**
+	 * Provide manager securty group if the manager is supporting policy mapping and security group
```
Comment (arvindn05) [inline designs/manager_policy_mapping/policy_mapping.md]: typo securty 

Provides manager Security group Id

File: designs/manager_policy_mapping/policy_mapping.md [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+	 */
+    String getManagerSecurityGroupId();
+
+	/**
+	 * Provide tag in case manager is not supporting security group
```
Comment (arvindn05) [inline designs/manager_policy_mapping/policy_mapping.md]: This is incorrect
tag is the manager policy id in case of IP/MAC
its the encapsulation tag in case of encapsulation

File: designs/manager_policy_mapping/policy_mapping.md [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+	 */
+    String getTag();
+}
+```
+
+`ManagerSecurityGroupInterfaceApi`: Add `SecurityGroupInterfaceElement` as a paramater to interface methods `createSecurityGroupInterface` and `updateSecurityGroupInterface`
```
Comment (arvindn05) [inline designs/manager_policy_mapping/policy_mapping.md]: please provide these interfaces as well.

**Comment by emanoelxavier** [comment]
@arvindn05 as discussed I am merging the doc.
